### PR TITLE
Variational Classifier

### PR DESCRIPTION
**Pull request by Pranav1006**
## Summary
Added a Variational Classifier, replacing the fusion head of the stsnet model with VariationalFusionHead. An extra switch was added to discriminator loss (implemented according to Dhuliawala's paper) to allow for ablation tests (just vc loss, just discriminator loss, both losses) and compare evaluation scores.

Additionally, some refactored SLURM scripts were added to be used for running train_eval on longleaf (may need more debugging on longleaf later). A fast-test flag was also added to allow for quick runs of the code, which sets iterations and subjects to 1, epochs to 2, and windows to 5.

## Future Steps
Next steps are to get graphs for the outputs of the new Variational Classifier, and record training logs for the different ablations.

### CODE REVIEW

File: src/eegproc/supervised/stsnet/data_representation.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+    for b_idx, band in enumerate(bands):
+        bsig = band_signals[band]
+        for t in range(n_windows):
+            start = t * window_size
+            end   = start + window_size
+            if end > bsig.shape[-1]:
+                break
+            result[t, b_idx] = compute_spd(bsig[:, start:end])
+
```
[the remark this answers is not in the document]
Comment (Pranav1006) [reply src/eegproc/supervised/stsnet/data_representation.py:207]: @copilot apply changes based on [this feedback](https://github.com/EEGo-UNC/EEGProc/pull/24#discussion_r3231358258)

File: src/eegproc/supervised/stsnet/README.md [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+```
+stsnet/
+├── data_representation.py   # EEG → 4-D SPD tensor (ManifoldNet) & flattened sequence (BiLSTM)
+├── manifold_net.py          # ManifoldNet sub-model  (wFM layers + invariant layer → MO)
+├── prepare_datasets.py      # Convert raw DEAP and DREAMER dataset files into the NumPy format
+├── stsnet.py                # BiLSTMNet + FusionHead + full STSNet model + joint training
+├── train_eval.py            # LOSOCV experiment runner (DEAP / DREAMER)
+├── tests.py                 # Unit & smoke tests
+└── README.md
+```
```
[the remark this answers is not in the document]
Comment (Pranav1006) [reply src/eegproc/supervised/stsnet/README.md:26]: @copilot apply changes based on [this feedback](https://github.com/EEGo-UNC/EEGProc/pull/24#discussion_r3231358358)

File: src/eegproc/supervised/stsnet/README.md
Comment (VitorInserra) [inline src/eegproc/supervised/stsnet/README.md:3]: Missing: adapted to support variational learning objectives.

Comment (VitorInserra) [inline src/eegproc/supervised/stsnet/README.md]: Remove docker dependence

File: src/eegproc/supervised/stsnet/stsnet.py [hunk 3]
```diff
@@ -91,71 +92,194 @@ def call(self, x: tf.Tensor, training: bool = False) -> tf.Tensor:
...
+        cfg = super().get_config()
+        cfg.update({"n_classes": self.n_classes})
+        return cfg
+
+
+# ---------------------------------------------------------------------------
+# Variational Fusion Head
+# ---------------------------------------------------------------------------
+
+class VariationalFusionHead(tf.keras.layers.Layer):
```
Comment (VitorInserra) [inline src/eegproc/supervised/stsnet/stsnet.py]: Could we rewrite the variational head outside of the stsnet scope? We will have to rewrite it later if not (or things can get ugly).

File: src/eegproc/supervised/stsnet/stsnet.py [hunk 4]
```diff
@@ -91,71 +92,194 @@ def call(self, x: tf.Tensor, training: bool = False) -> tf.Tensor:
...
+
+        # Term 3: lambda_ * KL(p(y) || p_pi(y))
+        p_y         = tf.reduce_mean(y_onehot, axis=0)
+        log_p_pi    = tf.nn.log_softmax(self.log_class_prior)
+        kl_class_prior = tf.reduce_sum(p_y * (tf.math.log(p_y + 1e-8) - log_p_pi))
+        prior_term  = lambda_ * kl_class_prior
+
+        return xent + kl_term + prior_term
+
+    def discriminator_loss(self, mh: tf.Tensor, y: tf.Tensor) -> tf.Tensor:
```
Comment (VitorInserra) [inline src/eegproc/supervised/stsnet/stsnet.py]: Nice!!! Only missing closed-form gaussian when "vc_only".

File: src/eegproc/supervised/stsnet/stsnet.py [hunk 5]
```diff
@@ -91,71 +92,194 @@ def call(self, x: tf.Tensor, training: bool = False) -> tf.Tensor:
...
+        mh      : (batch, d)
+        y       : (batch,) int32
+        beta    : KL weight for encoder/prior alignment  (0 = disabled)
+        lambda_ : KL weight for class-prior alignment    (0 = disabled)
+        """
+        y_onehot = tf.one_hot(y, self.n_classes)
+
+        # Term 1: cross-entropy (always active)
+        logits = self(mh, training=True)
+        xent = tf.reduce_mean(
```
Comment (VitorInserra) [inline src/eegproc/supervised/stsnet/stsnet.py]: Add hyperparameter to this term. Leave it continuous but in ablations we should just run 0 or 1.

File: src/eegproc/supervised/stsnet/tests.py
Comment (VitorInserra) [inline src/eegproc/supervised/stsnet/tests.py:1]: These tests are duplicates of the ones in `tests/test_stsnet.py`. Move all unit tests and smoke tests that have to be automated to `tests` directory and delete this if there are no extras.

**Review by VitorInserra — CHANGES_REQUESTED** [review]
Great work!! Last stretch!

**Review by VitorInserra — CHANGES_REQUESTED** [review]
`tests.py` is getting 3 failed tests

File: src/eegproc/supervised/stsnet/README.md
Comment (VitorInserra) [reply src/eegproc/supervised/stsnet/README.md:3]: Also correct the `python tests.py` has a relative import issue. Running with -m is usually better, something like `python -m supervised.stsnet.tests` will fix it (need to cd back first).

Comment (Pranav1006) [reply src/eegproc/supervised/stsnet/README.md:3]: Added the missing part to the README, and removed tests.py (the whole file was a duplicate of test_stsnet.py in the tests folder)

File: src/eegproc/supervised/stsnet/tests.py
Comment (Pranav1006) [reply src/eegproc/supervised/stsnet/tests.py:1]: Tests moved and file deleted!

File: src/eegproc/supervised/stsnet/README.md
Comment (Pranav1006) [reply src/eegproc/supervised/stsnet/README.md]: README has been updated, no longer says that a dev container is needed.

File: src/eegproc/supervised/stsnet/stsnet.py (hunk 3 above)
Comment (Pranav1006) [reply src/eegproc/supervised/stsnet/stsnet.py]: It's been moved out of stsnet and renamed to VariationalClassifier.

File: src/eegproc/supervised/stsnet/stsnet.py (hunk 5 above)
Comment (Pranav1006) [reply src/eegproc/supervised/stsnet/stsnet.py]: Hyperparameter alpha has been added. Like beta and lambda, it defaults to 1.0 and is multiplied by the cross-entropy term.

**Comment by Pranav1006** [comment]
## New Changes
Added new Gaussian loss, which accomplishes the same thing as the discriminator estimation term. Hyperparameters were added for ablation of vc_loss components. Some suggestions:

beta 1 vs gamma 0, beta 0 vs gamma 1 -> This would let us compare the new loss and the old discriminator prediction loss
beta 0, gamma 0, lambda_ 0 -> Just cross entropy, provides a baseline without any of the variational components.

Though alpha exists, setting it to 0 would disable the cross-entropy loss, so this shouldn't be done, but we can try other weights (0.25, 0.5, 0.75, etc.).

**Review by VitorInserra — APPROVED** [review]
